[CL-1599] Import of Matomo visit data

adessy opened the pull request:
The PR implements the logic to import visit data from Matomo into the back-end database. The import will run periodically for all tenants as a CRON task. However, the PR does not install the CRON task just yet, as we'll first test the import manually for a few tenants (to see how Matomo handles to the load).

adessy committed:
CL-1599 Add `get_last_visits_details` to Matomo client

adessy committed:
CL-1599 Add support for minTimestamp query parameter

adessy committed:
CL-1599 Add basic utils for locking resources

adessy committed:
CL-1599 Rework get_last_visits_details

Most of the query parameters for Live.getLastVisitsDetails are actually optional. (The API is too flexible for its good.) So, I reworked the signature to make it easy to specify only the relevant parameters.

adessy committed:
CL-1599 Add support for filter_sort_order query parameter

adessy committed:
CL-1599 Implement helper service to import matomo visit data

adessy committed:
CL-1599 Add que job to import matomo visit data

adessy committed:
CL-1599 Define CRON job to import matomo visit data

adessy committed:
CL-1599 Add comment

adessy committed:
CL-1599 Add doc for the concurrency control mechanism

adessy committed:
CL-1599 Raise error if a request fails while importing matomo data

We observed in the past that sometimes `raise_if_error` fails because the response do not have the expected format (e.g. CloudFront times out). So I reworked it along the way (and made it public).

adessy committed:
CL-1599 Rerecording VCR cassette after changes to Matomo client

adessy commented:
@James It's still a draft PR, but you can already take a look if you want. Still a draft because:
- it doesn't import visits' locales and projects
- there are almost no tests (and I didn't test manually either)

Other than that, everything is already there and shouldn't change much. Now, I'm gonna wait for the other related PRs to be released before wrapping it up.

adessy committed:
CL-1599 Fix typos

adessy committed:
Merge branch 'CL-1598-visitor-data-model' into CL-1599/import-matomo-data-cron

adessy committed:
Merge branch 'CL-1598-visitor-data-model' into CL-1599/import-matomo-data-cron

# Conflicts:
#	back/Gemfile.lock

adessy committed:
CL-1599 Adapt attribute extraction to latest changes in CL-1598

jamesspeake committed:
[CL-1598] Refactored referrer_type to add key and remove multiloc. Fixed a couple of Rubocop offenses

jamesspeake committed:
[CL-1598] Shortened create syntax

Co-authored-by: adessy <adrien@citizenlab.co>

jamesspeake committed:
[CL-1598] Review fixes - new lines and required fields

jamesspeake committed:
[CL-1598] Review fixes - unique index

jamesspeake committed:
[CL-1598] Populate referrer type dimensions on db:migrate and db:reset

jamesspeake committed:
[CL-1598] Updated seeding prior to moving to ee

jamesspeake committed:
[CL-1598] Moved seeding to a seed file in the ee repo

jamesspeake committed:
[CL-1598] Simplified date factory based on code review

jamesspeake committed:
[CL-1598] Test changes based on review

jamesspeake committed:
[CL-1598] Simplified based on code review

Co-authored-by: adessy <adrien@citizenlab.co>

adessy committed:
CL-1599 Import matomo visit projects and locales

- Create dummy model classes to allow batch in/upserts.
- Add unique indexes to join tables to avoid duplicates
- Adapt the MatomoDataImporter

adessy committed:
CL-1599 Reworking extraction of visit attributes

adessy committed:
Merge branch 'CL-1598-visitor-data-model' into CL-1599/import-matomo-data-cron

# Conflicts:
#	back/engines/commercial/analytics/lib/tasks/analytics_tasks.rake
#	back/engines/commercial/analytics/spec/acceptance/analytics_api_spec.rb
#	back/engines/commercial/analytics/spec/factories/dimension_dates.rb
#	back/engines/commercial/analytics/spec/models/analytics/dimension_date_spec.rb

adessy committed:
CL-1599 Add tests for lock_manager.rb

adessy committed:
CL-1599 Adapt attribute extraction to latest changes in CL-1598

adessy committed:
CL-1599 Test import_latest_matomo_data_job.rb + fixes

adessy committed:
CL-1599 Fix get_last_visits_details

Both GET and POST are possible. But if the authorization token is in the body, it must be a POST.

adessy committed:
CL-1599 Fix import of Matomo visits

Return early if there are no projects or locales associated with a visit because `insert_all` raises an exception if it's called with an empty array.

adessy committed:
CL-1599 Convert empty strings to nil in visit attributes

adessy committed:
CL-1599 Move voodoo constant to Ruby constant

We move the `900` to a constant, so that we don't have to hardcode it in tests. We could maybe make this a parameter in the future if needed.

adessy committed:
CL-1599 Fix import_data

from_timestamp was set to 0 when `FactVisit.maximum(:matomo_last_action_time)` returned nil.

adessy committed:
CL-1599 Rework and test MatomoDataImporter

The previous implementation was making one too many request too matomo. It was requesting an extra batch of visits that was discarded.

adessy committed:
CL-1599 Add visitor_ids in Matomo fixtures

adessy committed:
CL-1599 Test and fix matomo_data_importer.rb

adessy committed:
Merge branch 'CL-1598-visitor-data-model' into CL-1599/import-matomo-data-cron

# Conflicts:
#	back/db/schema.rb

jamesspeake commented on back/config/schedule.rb:
```diff
@@ -0,0 +1,6 @@
+# Use this file to easily define all of your cron jobs.
+# Learn more: http://github.com/javan/whenever
+
+every 6.hours do
```
Does this only get run on one application server?

adessy replied:
Yes! well actually on zero atm :) but the CRON task will be installed only on the first server of each cluster.

jamesspeake commented on back/db/migrate/20220929125457_add_index_to_analytics_dimension_projects_fact_visits.analytics.rb:
```diff
@@ -0,0 +1,13 @@
+      :analytics_dimension_projects_fact_visits,
+      %i[dimension_project_id fact_visit_id],
+      unique: true,
+      name: 'i_analytics_dim_projects_fact_visits_on_project_and_visit_ids'
```
Not really an issue - and possibly my way is wrong - but this index name is inconsistent with with shorter way I've named the indexes on the same table

adessy replied:
Ah yeah 😐 I don't think we have a convention for that.

jamesspeake commented on back/engines/commercial/analytics/app/jobs/analytics/import_latest_matomo_data_job.rb:
```diff
@@ -0,0 +1,74 @@
+    # multiple locks (`MAX_CONCURRENCY`). Each job is pseudo-randomly assigned to one of
+    # these locks (based on the hash of its target site id). If it fails to acquire this
+    # lock, the job is rescheduled (even if some other locks are available).
+    MAX_CONCURRENCY = 10
```
Is this something we should be able to change in an environment variable? Would that make it easier to change if there are issues?

adessy replied:
Yes, it would probably be a bit easier. Will change.

jamesspeake commented on back/engines/commercial/analytics/app/jobs/analytics/import_latest_matomo_data_job.rb:
```diff
@@ -0,0 +1,74 @@
+
+    def run(tenant_id, min_duration: 1.day, max_nb_batches: 5, batch_size: 250)
+      Tenant.find(tenant_id).switch do
+        matomo_site_id = AppConfiguration.instance.settings.dig('matomo', 'tenant_site_id')
```
Should we do a check here before doing anything else that the Matomo site_id exists and also isn't the default tenant_site_id?

adessy replied:
Not sure. I mean we could, but I think a better solution would be to:
1. prevent this from happening again (make tenant creation fail when matomo configuration fail)
2. fix the misconfigured tenants.

jamesspeake replied:
Yes we should do both. There are currently a few issues with some recently created sites (identified today), so in order to get this out without worrying about that issue might be worth a quick belt and braces check here that matomo_site_id is not nil or 28.

jamesspeake commented on back/lib/citizen_lab/lock_manager.rb:
```diff
@@ -0,0 +1,22 @@
+module CitizenLab
+  # Implements a (binary) lock on the top of PostgreSQL advisory locks.
+  # Inspired from: https://www.kostolansky.sk/posts/postgresql-advisory-locks/
+  module LockManager
```
Probably just want to discuss what would happen if we got errors running code that has created a block. Does that lock get released automatically or do we end up with everything piling up behind it?

adessy replied:
The lock is also released if the transaction is rolled back.

adessy committed:
CL-1599 Rework VCR configuration

The VCR configuration is global, so I moved it to spec_helper.rb.

Unfortunately, it seems that VCR does not support a convenient way to dynamically adapt the cassette library directory (for example, to save the cassettes in the corresponding engine folder). So for now, we'll just change the cassette directory in the spec files before running the tests.

adessy committed:
CL-1599 Re-record vcr cassette for matomo client_spec.rb

jamesspeake reviewed (commented):
Done an initial review. In general, this is looking really good, but added a few questions.

adessy committed:
CL-1599 Add missing import

adessy committed:
CL-1599 Fix tests

Removing dependencies to environment (variables 'MATOMO_HOST' and 'MATOMO_AUTHORIZATION_TOKEN')

adessy committed:
CL-1599 Fix rubocop offenses

adessy committed:
CL-1599 Fix rubocop offenses

adessy committed:
CL-1599 Fix tests

Removing dependencies to environment (variables 'MATOMO_HOST' and 'MATOMO_AUTHORIZATION_TOKEN')

adessy committed:
CL-1599 Allow to set max concurrency for matomo import via environment

adessy committed:
CL-1599 Cancel import if matomo site is not configured

adessy committed:
CL-1599 Fix typo [skip ci]

jamesspeake commented on back/engines/commercial/analytics/app/jobs/analytics/import_latest_matomo_data_job.rb:
```diff
@@ -23,6 +23,8 @@ class ImportLatestMatomoDataJob < ::ApplicationJob
       Tenant.find(tenant_id).switch do
         matomo_site_id = AppConfiguration.instance.settings.dig('matomo', 'tenant_site_id')
+        return if matomo_site_id.blank?
+
```
We also should ignore if it's still set to the default 'fake tenant' ID of 28

adessy replied:
--You didn't see my other comment?--

Oops, that's me who didn't see yours :)

adessy committed:
CL-1599 Raise an error if matomo site id is misconfigured

Raise error if matomo site id is either an empty string, or the default matomo site id. There should probably not be a default value.

adessy committed:
CL-1599 Fix import_latest_matomo_data_job_spec.rb

adessy committed:
CL-1599 Fix flaky VCR tests

The configuration of the cassette library directory was dependent on the loading order of files.

jamesspeake commented on back/config/schedule.rb:
```diff
@@ -0,0 +1,8 @@
+# Use this file to easily define all of your cron jobs.
+# Learn more: http://github.com/javan/whenever
+
+every 6.hours do
```
Just a minor point. Wonder if making this configurable through .env would be an idea? Just in case we need to experiment with timings.

adessy replied:
Actually, I just realized that the automatic setup of cron tasks cannot work that way. So I'll leave things as they are for now. In the worst case, we'll just do it as we always did, by manually editing the crontab on the first node of each cluster.

nTraum replied:
We also have regular unix cron jobs running on every single cluster with certain rake tasks. should we maybe unify our approach to not have both systems running? They seem to be do the very same thing as far as I can see.

adessy replied:
Yes, this is what I was referring to by "manually modify the crontab on the first node of each cluster". It was an attempt to make things more automatic/reproducible, but that's more work than I expected. So I'll probably give up... for now. 

But curious, how did you end here? 😁🧐

nTraum replied:
Solely by accident and curiosity :)

jamesspeake reviewed (approved):
One minor comment, but other than that, good to go I think.

adessy committed:
Merge branch 'master' into CL-1599/import-matomo-data-cron

# Conflicts:
#	back/db/schema.rb
